### PR TITLE
elf: implements extended section header

### PR DESCRIPTION
**Pull request by m4b**
Based off of related #77 

1. using @zhuowei work, add support for sections greater than 65536
2. fix alignment issues in faerie when writing symtab, relocs, and section headers

### CODE REVIEW

**Comment by m4b** [comment]
@zhuowei feel free whenever you have time (or not), but could you let me know if this branch fixes your issue?

**Comment by zhuowei** [comment]
@m4b Thank you so much! I'll try applying this to Lucet and see if it works there.

File: src/elf.rs [hunk 1]
```diff
@@ -719,8 +727,81 @@ impl<'a> Elf<'a> {
...
+        // the offset is the head of how many program bits we've added
+        section.sh_offset = self.sizeof_bits as u64;
+        self.sections.insert(
+            idx,
+            SectionInfo {
+                header: section,
+                symbol: section_symbol,
+                name: idx,
+            },
+        );
```
Comment (philipc) [inline src/elf.rs]: I think it would be better if we stored this section separately (i.e. just a `Section`, no section symbol).

File: src/elf.rs [hunk 2]
```diff
@@ -719,8 +727,81 @@ impl<'a> Elf<'a> {
             self.nsections += 1;
         }
     }
+    pub fn add_extended_symtab(&mut self) -> usize {
+        // copypasta'd from add_progbits
+        let name = ".symtab_shndxr".to_string();
```
[the remark this answers is not in the document]
Comment (zhuowei) [reply src/elf.rs]: Wait, [LLVM got this section name wrong](https://github.com/llvm/llvm-project/commit/88838d5be75ae211ef5cbe47044eb4bb50c9fad6)?! Childhood ruined
I only looked at what Clang outputs, so I didn't know they had a misspelling. Good catch!

**Comment by jedisct1** [comment]
I merged these changes to Lucet in order to give them a try, and here is the verdict:

SUCCESS.

At least, the original test case that defines 0x10000 functions is now producing a valid ELF library, that Lucet can load and use just fine.

**Comment by m4b** [comment]
Does anyone want to pick the remainder of this up ?

Afaics:

1. need to use scroll in writing to extended section header vec.
2. Fix typos (basically all the things that Philipc mentions)
3. Evaluate remaining unaligned writeouts. I did it very quickly; also my align implementation adds sizeof T unnecessary bytes when the offset is already aligned (not a big deal but should be aware).

File: src/elf.rs [hunk 3]
```diff
@@ -719,8 +727,81 @@ impl<'a> Elf<'a> {
...
+
+        let final_symbol_count =
+            self.special_symbols.len() + self.sections.len() + self.symbols.len() + 1;
+        let section = SectionBuilder::new((final_symbol_count * 4) as u64)
+            .section_type(SectionType::SymTabShndx);
+
+        let mut section = section.name_offset(offset).create(&self.ctx);
+        section.sh_link = 2;
+        // the offset is the head of how many program bits we've added
+        section.sh_offset = self.sizeof_bits as u64;
```
[the remark this answers is not in the document]
Comment (m4b) [reply src/elf.rs]: i briefly looked at this and I didn't write any of this code so not super familiar with it :P so not sure exactly what its doing and also I'm tired haha.
regardless, if its true the `section.sh_offset` here must be aligned, then `self.sizeof_bits`, which if i recall correctly, is the size of the relocatable program bytes? This is also the offset at which the string table starts, which definitely doesn't need to be aligned, so I'm a little confused.
Can you or @pchickey, who according to git blame wrote this, clarify what should be done here, how to fix it?

[the remark this answers is not in the document]
Comment (philipc) [reply src/elf.rs]: I don't know who wrote this, but I think that's irrelevant. It's copy-paste code that is being used for a new purpose, but not taking into account the new requirements. This ties in with my suggestion below to store this section separately from the others. I think this should be written quite differently. I'll make an attempt at that in the next few days.

[the remark this answers is not in the document]
Comment (m4b) [reply src/elf.rs]: Ah, sorry; I thought your comment was on the other progbits. Yes we should remove the copy paste if possible, open to any suggestions. Feel free to push onto this PR if you like :)

I think should be able to, let me check

**Comment by philipc** [comment]
Added commit to:
- don't add a symbol for the .symtab_shndx section
- generate the symtab_shndx data as we write symtab
- add .symtab_shndx alignment (although this is probably unneeded because it comes after symtab now)
- using correct alignment for other sections
- avoid unnecessary alignment bytes

File: src/elf.rs [hunk 4]
```diff
@@ -814,14 +760,19 @@ impl<'a> Elf<'a> {
...
-        Self::align(&mut reloc_offset, sizeof_relocation);
+        let symtab_align = self.ctx.size() as u64;
+        Self::align(&mut symtab_offset, symtab_align);
+        let mut symtab_shndx_offset = symtab_offset + sizeof_symtab as u64;
+        let symtab_shndx_align = 4;
+        if sizeof_symtab_shndx > 0 {
+            Self::align(&mut symtab_shndx_offset, symtab_shndx_align);
+        }
+        let mut reloc_offset = symtab_shndx_offset + sizeof_symtab_shndx;
+        let reloc_align = self.ctx.size() as u64;
```
Comment (m4b) [inline src/elf.rs]: I'm confused, how could this be the same as `Relocation::size(true, self.ctx)`? Also the naming is somewhat inconsistent; the function calls it `sizeof_t`, but it's called align here.

File: src/elf.rs [hunk 5]
```diff
@@ -814,14 +760,19 @@ impl<'a> Elf<'a> {
...
+        Self::align(&mut reloc_offset, reloc_align);
         let mut sh_offset = reloc_offset + sizeof_relocs as u64;
-        let sizeof_shdr = Section::size(&self.ctx) as u64;
-        Self::align(&mut sh_offset, sizeof_shdr);
+        let shdr_align = self.ctx.size() as u64;
```
Comment (m4b) [inline src/elf.rs]: Similarly, what method is being called here? Why was this changed from `Section::size` ? Even if they are identical, it is less readable imho.

File: src/elf.rs [hunk 6]
```diff
@@ -814,14 +760,19 @@ impl<'a> Elf<'a> {
 
         // alignment required for below
         let mut symtab_offset = strtab_offset + self.sizeof_strtab as u64;
-        let sizeof_symbol = Symbol::size(self.ctx.container) as u64;
-        Self::align(&mut symtab_offset, sizeof_symbol);
-        let mut reloc_offset = symtab_offset + sizeof_symtab as u64;
-        let sizeof_relocation = Relocation::size(true, self.ctx) as u64;
-        Self::align(&mut reloc_offset, sizeof_relocation);
+        let symtab_align = self.ctx.size() as u64;
```
Comment (m4b) [inline src/elf.rs]: Ditto here, this is less readable than explicit `Symbol::size`; I want to know which sizeof I'm getting

File: src/elf.rs [hunk 7]
```diff
@@ -727,83 +725,31 @@ impl<'a> Elf<'a> {
...
-            self.add_extended_symtab();
-        }
-
         /////////////////////////////////////
         // Compute Offsets
         /////////////////////////////////////
-        let sizeof_symtab = (self.symbols.len() + self.special_symbols.len() + self.sections.len())
-            * Symbol::size(self.ctx.container);
+        let symbol_count = self.symbols.len() + self.special_symbols.len() + self.sections.len();
+        let sizeof_symtab = symbol_count * Symbol::size(self.ctx.container);
```
Comment (m4b) [inline src/elf.rs]: I'm really confused why these `X::size` calls changed; why use this here, and not below? Please change them back to their explicit versions.

File: src/elf.rs (hunk 4 above)
Comment (philipc) [reply src/elf.rs]: It's not the same, that's why I changed it. The structures aren't aligned to their size, they're aligned to the largest alignment required by their fields, which is an address, and `self.ctx.size()` gives the size of an address.

**Comment by philipc** [comment]
> I really would like to see the X::size calls changed back.

This change is what I listed as: using correct alignment for other sections. The X::size calls were wrong.

> We should merge this, but I'll say I do think I liked the optional section header version better from the perspective that the section header extended table is optional, insofar as it doesn't have to exist.

It's not a 'section header extended table'. It's an extended symtab, contains the section header indices for the symtab entries. It makes perfect sense to me for it to be generated at the same time as the rest of the symtab.

File: src/elf.rs (hunk 4 above)
Comment (philipc) [reply src/elf.rs]: Also, this matches the alignment that is set for the relocation sections, etc: https://github.com/m4b/faerie/blob/49aac0ca5b49c6ef95a57a95d8a006ab73e998e5/src/elf.rs#L304 (well, that should be using ctx.size() too, but hopefully you see what I mean).

Comment (m4b) [reply src/elf.rs]: Yes you're totally right, sorry for the confusion - very good catch! Probably the math turns out to yield a an address which is also aligned to a pointer width, but we should be correct, thank you!

Just a note: I didn't think the alignment is aligned to the largest alignment required by their fields, as far as I understood, it's the alignment required to make arbitrary acceses safe w.r.t. an aligned pointer, e.g., this formula must be true of for a correctly aligned `address`:
```
(address + (sizeof_t * i)) mod sizeof_ptr = 0
```
e.g., when our starting address is aligned, offsetting by index `i * sizeof foo` is guaranteed to yield an aligned pointer:
```
address: [ foo; foo; foo; ]
  aligned ^
                       ^ (address + sizeof_foo) mod sizeof_ptr = 0
```
Maybe it's the same thing to say aligned to largest alignment required by fields, but I've always understood it as above, but perhaps that's wrong?

**Comment by m4b** [comment]
I might push a test for this later, then I think we are ready to merge? May need the hotfix in goblin to properly test though?

File: src/elf.rs (hunk 4 above)
Comment (philipc) [reply src/elf.rs]: > it's the alignment required to make arbitrary acceses safe w.r.t. an aligned pointer

Yes that's true, but generally the way you do it that is if `align` is the alignment required for a field within a structure, then you layout the structure so that the field is placed at an `offset` within the structure which satisfies `offset % align == 0`. Given that, it is then enough to ensure that the structure is placed at an address such that `address % align == 0` for every field's alignment, which is equivalent to ensuring that the address uses the largest alignment required by fields.

The other thing that is done is that padding is added to the end of the structure to ensure that `size % align == 0`. This allows arrays of the structure to work automatically.

File: src/elf.rs [hunk 8]
```diff
@@ -903,24 +854,40 @@ impl<'a> Elf<'a> {
...
+        } else {
+            Vec::new()
+        };
+        let mut offset = 0;
         file.seek(Start(symtab_offset))?;
         for symbol in self.special_symbols.into_iter() {
             debug!("Special Symbol: {:?}", symbol);
             // the special symbols's section indexs have special meanings
             // so we don't do the shn conversion here
+            if sizeof_symtab_shndx > 0 {
```
[the remark this answers is not in the document]
Comment (philipc) [reply src/elf.rs]: I disagree, but this is minor so I've changed it back.

File: src/elf.rs [hunk 9]
```diff
@@ -750,7 +789,11 @@ impl<'a> Elf<'a> {
         header.e_machine = machine.0;
         header.e_type = header::ET_REL;
         header.e_shoff = sh_offset;
-        header.e_shnum = self.nsections;
+        header.e_shnum = if self.nsections >= SHN_LORESERVE.into() {
```
Comment (m4b) [inline src/elf.rs]: was hoping this would have used the need boolean, but it's not too important. going to push a test and merge this :)

Comment (philipc) [reply src/elf.rs]: This is different though. This is whether the number of sections is too large, the boolean is whether the symtab references a large section index. Often these will be the same, but it is not required.

Comment (m4b) [reply src/elf.rs]: What do you mean? `need_symtab_index = self.nsections >= SHN_LORESERVE.into()`

Comment (philipc) [reply src/elf.rs]: Yes, but that test is not exact, and there's a comment above it explaining. There's potential to refactor that in future, and keeping these checks separate (as they logically are) makes it less fragile for refactors.

**Comment by m4b** [comment]
Well this was epic and awesome, thanks for everyone help as usual, and thanks to new contributor @zhuowei :) Once CI is green, will rebase and merge, without squashing so authorship is preserved better :)

**Comment by zhuowei** [comment]
@m4b @philipc I also tested this in Lucet and it now generates a working .so from a WASM with many functions.

Thank you so much for your work on this, and for all your work on faerie!

I'm OK with squash-merge - just a credit/link to the issue in the commit message's probably fine.

**Comment by m4b** [comment]
I squashed, because it makes history better usually, and is less prone to breaking `git bisect` :) Thanks for everyone's work here!
